src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,235 @@
+/**
+ * @file multihead_attention_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the MultiheadAttention class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "multihead_attention.hpp"
+
+#include "../visitor/load_output_parameter_visitor.hpp"
+#include "../visitor/save_output_parameter_visitor.hpp"
+#include "../visitor/backward_visitor.hpp"
+#include "../visitor/forward_visitor.hpp"
+#include "../visitor/gradient_set_visitor.hpp"
+#include "../visitor/gradient_update_visitor.hpp"
+#include "../visitor/gradient_visitor.hpp"
+#include "../init_rules/glorot_init.hpp"
+#include "softmax.hpp"
+#include "linear.hpp"
+
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType>
+MultiheadAttention<InputDataType, OutputDataType>::MultiheadAttention()
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+MultiheadAttention<InputDataType, OutputDataType>::MultiheadAttention(
+    const size_t embedDim,
+    const size_t numHeads,
+    const float dropout,
+    const bool addZeroAttn,
+    const size_t kdim,
+    const size_t vdim) :
+    embedDim(embedDim),
+    numHeads(numHeads),
+    dropout(dropout),
+    addBiasKV(addBiasKV),
+    addZeroAttn(addZeroAttn)
+{
+  typedef typename InputDataType::elem_type ElemType;
+
+  this->kdim = kdim ? kdim : embedDim;
+  this->vdim = vdim ? vdim : embedDim;
+  this->headDim = embedDim / numHeads;
+
+  Log::Assert(this->headDim * numHeads == this->embedDim,
+      "`embedDim` must be divisible by `numHeads`.");
+
+  qWeight.set_size(embedDim, embedDim);
+  kWeight.set_size(embedDim, this->kdim);
+  vWeight.set_size(embedDim, this->vdim);
+  outWeight.set_size(embedDim, embedDim);
+
+  qBias.set_size(embedDim, 1);
+  kBias.set_size(this->kdim, 1);
+  vBias.set_size(this->vdim, 1);
+  outBias.set_size(embedDim, 1);
+
+  queryLinear = new Linear<>(embedDim, embedDim);
+  keyLinear = new Linear<>(this->kdim, embedDim);
+  valueLinear = new Linear<>(this->vdim, embedDim);
+  softmaxModule = new Softmax<>();
+  dropoutModule = new Dropout<>(dropout);
+  outLinear = new Linear<>(embedDim, embedDim);
```

Comment:
I think it should be `Linear<InputDataType, OutputDataType>`. The same rule applies to the lines above.

Reply:
Oh! completely overlooked this thing. Thanks.